hardware/max_storage_temp_tutorial.ipynb

```diff
@@ -2403,279 +3444,2136 @@
79av8/N4rsV31745bT8JYhu/098Vf/ljjLv0/e6c5HLK9mzKiufPXbn66UFj15jKrVWw4VuStjbbvdLQBf6rhcN//xe8stperlvd75g9/B0Hxu7/Fhampqamo6RvqKPyRJkmihUuoHw1uQabPy7Ible6JF0ktt/+WPjzgGHwHSl9tv9eP5Q5K0zexiKUmSUhep2uDcmgda0gztWrm6eRwfdLEbe1IJrWsex4ckSeuxBZkkSZIkSZL22lemzoAkSZIkSZI0JQNkkiRJkiRJ2msGyCRJkiRJkrTXDJBJkiRJkiRprxkgkyRJkiRJ0l4zQCZJkiRJkqS9ZoBMkiRJkiRJe80AmSRJkiRJkvaaATJJkiRJkiTtNQNkkiRJkiRJ2msGyCRJkiRJkrTXDJBJkiRJkiRprxkgkyRJkiRJ0l774akzIEldXr169erg4PPnqfMh9Xf46vPnz58PDn5/6oxIkiRJ6sEAmaQtc/hq6hxIdc8GciVJkqQtZIBM0taxZY7mpLRw/ONTZ0OSJEnSGhyDTJIkSZIkSXvNAJkGefXq1atXr96+LenR0WbWf3xc0o8fS/r5c0kfH/n/1Pu7aj7L+4eHcb1nZ+3LHR2x3EuVZ9f32+//Z2dxucPDzXwv3duJ5f3+fdx/yuPiYuz8jbufi8dB3P/8vfJ6+O9k+faGl2P8flj+06e5l7skSZKkPfT5c+msZLqfKUdBV9q+/Onp8vWenvZb/9u38XMfP5b0/fu4Hl6TcgPfnf/18rO4v33z2V4uBAZ45+PHrvJb7/vtn8+u77ff/9ke71xcrHbcxXzx1+fPf3zgdt69i8sTkOT4+fSppATY+h5P/Y+/dfa77TiI+b66iu+TL/Z39d/JeuVIgI418H7+HR4dLT9++Ry/m3weYLvNeeFlz59lP79+9jc/U2oHBwcH3/x//9sjb4dyzvtPeni42f18uXJd9/uYa77nlp++5TjO+v6T/4Tf8dT7Z2pqampqampaSyfPgOnEB0C6MR5r+dqN9vLl8w11c+Mc/88NIS2Y2M7VVfuNfk5r/28CCbX9HZrP9v18fMzbK+nZ2arlt2559v1+a/8n//H/7N/j4/L18L2xvzmAWN7nk9++/9WDYdshgFILoHE8UR614yXnk+Nv1d9Pv/2uB6z4i/xT/uwH5bL672Sccszf02L+lpdX87nadl/6vBm3f3dXSv8X/nmpHRwcHPyJ//o7n9uWWwxsUR58f/n44vtqAqHtv+N372r/7/v+8v2crpxXzeem8h2/nyYwmcs1Lze03OdSjmOtj+O063dvampqampqajpV6iD9+hK6TuX3Hx7KAfPmzaprpotV+38X11u29/QUX3/40L7801POd1n+4SFunxvPZn/b88N2apqufH3zGbskckN7eVnSHMC4v1+1nNfNZy6f9uOhJuef9OqK/c/5KGpdSJ+fSxqXf37+M5+HbYdyyNulHOrHddz/Dx/aj6uu8qr9fvrud60rM8cP6yEAg5ubWA6r/k4wvBzL+lh/E+iJ5bK4voiAwvX1wSyV/bu7+1/Cu1/5//x++eP/XPtc7hJc64pM4LV2nuR3zPGQcXzUjr83b+gqW17n8w/HY2N5128+v9h1vK/a9x275uYuuv/Zf1ZSugg3x3F5TWCG/SGwdXnZb3/4P5//X//Xkv7Tf1o+z3JcV774In6u+b2W5buuMzc37dtt/l/y37zf73vk97hYHsvLkeUo92a7fF/Lvx+Wv72N/5/r71qSJO0jxyDTl7x5Uyq6r17FdPXAGNrXS0qFPVf0mxvG2thIY+WnvMsNRRMwaLdqPrkheHqKN7Zs7+Ji3DG71i1Pblz4/msBkNwFjsAH+0cApDbmVHPcldcnJyUlIBUDKP/bxa8eDNtOvgHM+R9rTL2hv59h+72I/79+Hb+fJmC43v5k/csxHhfc6JNPXp+eLg/IrY/ASNf3Sz7ifuR08fPleyo3/P/Kf/l/Cf/7V+7/wxQq5Xvlezo/Lyktw25vS3p/n7//rvPCqmPORZyXrq9jgCoHmnJAP6sdd/xO+f6H/h+0tOX3Qnn+2T8b96M5jxMIi/vFcTp0f1jPv/PvsJ74+fv7HFjKgaxa+cf9J8CYv3e2Vzufdn2P7M/5efvxWCvHxcAY643HX+37yQHdlxtLVJIkqS8DZPoSBnNvblxjmivq9eXzmpevlxYiOdCRb0xyF62xsd36+tfLJzcY3JjRoo4n+Xm5vuVXu3FftzybFlP5hnN5fglIsH9dN3S1fHcFTvpuh4BRE4AsKeVRu/EbaujvZ9X9bva/pARY+H7Y364A21BDy5H9zV2icwBk3XJf1P67GH9SgLKdEshoWjYWf/Pz76Sl83723e8coMjbz110+1o3EM/5gQAM3ysthAj05RZCHLdd/39pXfvD77T2PXIe5XP8Lmq/w7z/LJdbNnP+zfkaq6XxUARym/zE85ckSdL2soulvqSrZVbfLmK15WtdGrkBzE+y7+7iDTjLbbpLRt8AXHc+v3wD3dyQvn7d3kKEG0OWzzcctfLjhqlWLpsuT/LLjVPT1SsGKpjtsH/Xqy8HAAiD/cjtf37w+fPnz8//5r/xqt922D+6IDKWFuVPi5B1AzVDfw/99ruu9r02+7Xe/mRDy/H5ueQrf65pWbbZG2sCRgQ46t9PDDB2tSDNynH2v7//enn5b5bkXzj/rz4fHBwc/IO7P/uD4zS32Mwt7Gr5urmJLdnyeYDyrn3fDw9xuxgeIOO8VdbH95v3h/ImYJy3S2Cn6/81tNBiLEew/3S17huQ77s/NU2LqLg+zke1/WH/c1fHHOjOLSDJ16rnbY43zhs5oF0rx+Z3VN7Pgb+my3v8fljuw4d4Xh76O5MkSdq8H3QBmjobklT36tWrV02I7PBVOWv9/tTZ0obFG/TFrnPLP0cAjxt7Ao4E5sZrgVOOzz/+g1fPHqdSixgQpAvo+C1IJUmSVmUXS0nSLH15jK8+LTtzy8HYFYwWd5sb80
```

Comment:
@SenWu I just noticed that this LF had been commended out.
Can you add an explanation or maybe just remove the LF completely?

Reply:
Sure, this LF didn't label any candidate in this tutorial.

Reply:
Does this actually cause some issue for LFAanalyzer?
If you haven't tested the impact on LFAnalyzer, I'll quickly test it.
Even if it is the case, the issue is from Snorkel (LFAnalyzer). Am I correct?

Reply:
If I remember correctly, the issue is as follow:

Assume you have 10 LFs and only 9 of them can actually label the candidates (1 of them only outputs ABSTAIN). 

1. After applying the LFs, your label matrix will be sparse matrix with dimension N*9 where N is the number of candidates and 9 is the 9 LFs that can label the candidates.
2. When you feed the LF names and label matrix you will have an issue about dimension miss match since there are 10 LF names in total and sparse matrix only have 9 there.

This will cause an issue.

Reply:
I see. I tested and got the following error:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-51-a25c3cb6eed2> in <module>
----> 1 LFAnalysis(L=L_train[0], lfs=sorted(stg_temp_lfs + stg_temp_lfs_2, key=lambda lf: lf.name)).lf_summary(Y=L_gold_train[0].reshape(-1))

~/.venv/lib/python3.7/site-packages/snorkel/labeling/analysis.py in __init__(self, L, lfs)
     44             if len(lfs) != self._L_sparse.shape[1]:
     45                 raise ValueError(
---> 46                     f"Number of LFs ({len(lfs)}) and number of "
     47                     f"LF matrix columns ({self._L_sparse.shape[1]}) are different"
     48                 )

ValueError: Number of LFs (17) and number of LF matrix columns (16) are different
```

Reply:
`labeler.get_keys()` instead of the sorted lfs works like below:
```
LFAnalysis(L=L_train[0], lfs=labeler.get_keys()).lf_summary(Y=L_gold_train[0].reshape(-1))
```

However, a user may wonder where their `LF_too_many_numbers_row` is silently gone from the label matrix. Also this is hacky because labeler.get_keys() returns `List[LabelKey]` while LFAnalysis expects lfs to be `List[LabelingFunction]`.

For now, I think a comment to explain "why `LF_too_many_numbers_row` should be not be included in the list of lfs" is just fine, but for a longer term, we should address this by e.g., giving a user a warning during `labeler.get_label_matrices` or `labeler.apply`.

Reply:
Agree! Let's add a comment in the PR?

Reply:
A comment was added by a0a93b33daecbd974d73114cd2cca2520e7d39a3.